### PR TITLE
[SPARK-38838][SQL] Support ALTER TABLE ALTER COLUMN commands with DEFAULT values

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

Support ALTER TABLE ALTER COLUMN commands with DEFAULT values.

For example:

```
CREATE TABLE t(i BOOLEAN, s STRING) USING PARQUET;
ALTER TABLE t ALTER COLUMN s SET DEFAULT CONCAT('abc', 'def');
INSERT INTO t VALUES(TRUE, DEFAULT);
SELECT s FROM t WHERE i = TRUE;
> "abcdef"
```

### Why are the changes needed?

This reduces work for users to write INSERT INTO commands into tables by allowing them to leave out some of the column values in reasonable ways.

It also helps table owners add columns with values to auto-generate useful information into the table contents, such as date or time values.

Does this PR introduce any user-facing change?
Yes, it allows ALTER TABLE ALTER COLUMN commands to specify optional DEFAULT values for the columns. These default values are stored inside the column metadata.

### How was this patch tested?

Unit test coverage was added as part of this PR.

### CODE REVIEW

**Comment by dtenedor** [comment]
@gengliangwang this is now ready for you to check out at your convenience. In addition to the ALTER TABLE ALTER COLUMN support, I also added a function to control which table types are permitted for ALTER TABLE ADD COLUMN. In this way, we can make sure that each data source supports returning the default expression when the corresponding value is not present in storage (instead of NULL).

**Comment by gengliangwang** [comment]
@dtenedor Got it. I was busy over the past few days. I will take a close look tomorrow.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 1]
```diff
@@ -353,7 +353,18 @@ case class AlterTableChangeColumnCommand(
...
         // Create a new column from the origin column with the new comment.
-        addComment(field, newColumn.getComment)
+        val withNewComment: StructField =
+          addComment(field, newColumn.getComment)
+        // Create a new column from the origin column with the new current default value.
+        if (newColumn.getCurrentDefaultValue().isDefined) {
+          if (newColumn.getCurrentDefaultValue().get.nonEmpty) {
+            addCurrentDefaultValue(withNewComment, newColumn.getCurrentDefaultValue())
+          } else {
+            withNewComment.clearCurrentDefaultValue()
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:363]: What if the column is string type and the default value is set as `""`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala [hunk 2]
```diff
@@ -45,49 +45,62 @@ import org.apache.spark.sql.types._
...
-case class ResolveDefaultColumns(
-  analyzer: Analyzer,
-  catalog: SessionCatalog) extends Rule[LogicalPlan] {
+case class ResolveDefaultColumns(analyzer: Analyzer) extends Rule[LogicalPlan] {
 
   // This field stores the enclosing INSERT INTO command, once we find one.
   var enclosingInsert: Option[InsertIntoStatement] = None
   // This field stores the schema of the target table of the above command.
   var insertTableSchemaWithoutPartitionColumns: Option[StructType] = None
+  // This field records if we've replaced an expression, useful for skipping unneeded copies.
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:55]: Is this change related to the alter column change? If not, let's open a separate PR for it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala (hunk 1 above)
Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:363]: Good question. I confirmed that the parser returns the quotes in the default value string, and that passing in a completely empty string is prohibited (since the parser expects a token there). I added a couple unit tests to the `DDLParserSuite` to cover this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -3837,22 +3837,26 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
...
-    if (action.defaultExpression != null) {
-      throw QueryParsingErrors.defaultColumnNotImplementedYetError(ctx)
-    }
-    if (action.dropDefault != null) {
-      throw QueryParsingErrors.defaultColumnNotImplementedYetError(ctx)
-    }
+    val setDefaultExpression =
+      if (action.defaultExpression != null) {
+        Option(action.defaultExpression()).map(visitDefaultExpression)
+      } else if (action.dropDefault != null) {
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:3843]: Good catch! I added a check for this with a unit test.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala (hunk 2 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:55]: This field replaces the usage of `Option[LogicalPlan]` in all the methods below. When adding more logic, I found that the method call sequence got very complicated when checking the optional values. This PR adds more methods to the sequence, so it is related :)

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:55]: Sorry I misunderstood your comment. I will prepare a separate refactoring-only PR for this change and send it out separately.

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:55]: This refactoring-only PR is now ready for review at: https://github.com/apache/spark/pull/36398.

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala:55]: Let's drop the changes of this file in this PR

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala]: SG, I integrated in the recent refactoring changes and pushed the result to this PR. This `ResolveDefaultColumns` file is no longer changed in this PR anymore.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 4]
```diff
@@ -1440,6 +1440,75 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
+        sql(insertDefaults)
+      }.getMessage.contains(Errors.COMMON_SUBSTRING))
+      // The default value has an explicit alias. It fails to evaluate when inlined into the VALUES
+      // list at the INSERT INTO time.
+      sql("alter table t alter column s set default (select 42 as alias)")
+      assert(intercept[AnalysisException] {
+        sql(insertDefaults)
+      }.getMessage.contains(Errors.BAD_SUBQUERY))
+      // The default value parses but the type is not coercible.
+      sql("alter table t alter column s set default false")
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1494]: Shall we have a follow-up to throw an exception on this alter command?

Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1494]: Shall we have a follow-up to throw an exception on this alter command?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 5]
```diff
@@ -1440,6 +1440,75 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
+      assert(intercept[AnalysisException] {
+        sql(insertDefaults)
+      }.getMessage.contains("provided a value of incompatible type"))
+      // The default value is disabled per configuration.
+      withSQLConf(SQLConf.ENABLE_DEFAULT_COLUMNS.key -> "false") {
+        sql("alter table t alter column s set default 41 + 1")
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1500]: It seems this is working?

Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1500]: It seems this is working?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 6]
```diff
@@ -1440,6 +1440,75 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
...
+        Seq(
+          Row(true, null, null),
+          Row(true, "abcdef", null),
+          Row(true, "abcdef", 42),
+          Row(true, "abcdef", null)
+        ))
+    }
+  }
+
+  test("SPARK-38338 INSERT INTO with defaults set by ALTER TABLE ALTER COLUMN: negative tests") {
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1467]: wrong jira number. It should be 38838

**Comment by gengliangwang** [comment]
LGTM overall. @dtenedor Could you take a look at the test failures?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 5 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1500]: Probably we should define the scope of the conf `SQLConf.ENABLE_DEFAULT_COLUMNS`

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala (hunk 6 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1467]: Good catch, thanks!

**Review by dtenedor — COMMENTED** [review]
Thanks for your review!

**Comment by dtenedor** [comment]
> LGTM overall. @dtenedor Could you take a look at the test failures?

@gengliangwang SG, this is done. ✔️

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master
